Fix y-boundaries of Coordinates fields

johnomotani opened the pull request:
Make fields in `Coordinates` smooth at the y-boundaries. Previously the y-boundary guard cells were set equal to the nearest grid cell, so gradients of metric components (and therefore fields like `G1`, `G3` that are used in `Delp2` and Laplace inversion) could have large spikes at the y-boundaries.

This PR:
* loads y-boundary guard cells from grid files, if they are available
  * these grids can be generated by Hypnotoad as upgraded in #1647
* if not available in the grid file, extrapolates smoothly (3rd order accurate) into the boundary guard cells
  * handles the case of grid files without x-boundary guard cells, more for the sake of symmetry between x- and y-boundaries which makes the code easier to follow than any practical need for the feature.
* fixes a string-comparison bug in the check of `coordinates_type` in the grid file
* writes the `ny_inner` variable that specifies the y-index position of the 'upper' divertor targets in double null grids from `BoutMesh::outputVars()`

ZedThree commented on src/mesh/parallel/fci.cxx:
```diff
@@ -327,7 +327,7 @@ Field3D FCIMap::integrate(Field3D &f) const {
   std::string coordinates_type = "";
   if (!mesh.get(coordinates_type, "coordinates_type")) {
-    if (coordinates_type != "fci") {
+    if (strcmp(coordinates_type.c_str(), "fci") != 0) {
```
Why this change? `coordinates_type` is a `std::string`, so comparison to a literal is fine.

johnomotani replied:
This was a strange one. Somehow between Hypnotoad and netCDF, the `coordinates_type` string has a length that is 1 longer than it should be - e.g. here coordinates_type would have length 4 - so that the comparison to a string literal fails. I think the string just has an extra null byte at the end, because any normal way of printing it only shows the actual characters. I can't see a bug in our code, neither in `tools/idllib/file_write_attribute.pro` or `src/fileio/impls/netcdf4/ncxx4.cxx` - we never set the length explicitly. So this work-around seems necessary :-(

johnomotani replied:
It does deserve a comment to explain why its doing such an ugly string comparison though...

johnomotani commented on src/mesh/coordinates.cxx:
```diff
@@ -17,6 +17,127 @@
+          result(bndry->x, bndry->y) =
+              (9. * (f(bndry->x - bndry->bx, bndry->y - bndry->by) + f(bndry->x, bndry->y))
+               - f(bndry->x - 2 * bndry->bx, bndry->y - 2 * bndry->by)
+               - f(bndry->x + bndry->bx, bndry->y + bndry->by))
```
It does require at least 2 guard cells, the same as `interp_to` so this should always be true for a direction in which some field is staggered, but I'll add ASSERTs to check. In the direction into the grid, it's OK because it only needs 2 points inside the boundary; there must be at least one grid point and one guard cell (since if there are no guard cells there is no boundary cell to fill), so these two points are always available (if this expression is evaluated at all).

johnomotani commented on src/mesh/coordinates.cxx:
```diff
@@ -83,34 +223,43 @@ Coordinates::Coordinates(Mesh *mesh)
-        mesh->sourceHasVar("g_13") and mesh->sourceHasVar("g_23")) {
+    if (mesh->sourceHasVar("g_11") and mesh->sourceHasVar("g_22")
+        and mesh->sourceHasVar("g_33") and mesh->sourceHasVar("g_12")
+        and mesh->sourceHasVar("g_13") and mesh->sourceHasVar("g_23")) {
```
Good point. We did `any_of`/`all_of` in #1564 which only looked at the staggered `Coordinates` constructor, not this one. I'll copy the strategy here :+1:

johnomotani commented on src/mesh/data/gridfromfile.cxx:
```diff
@@ -227,149 +267,122 @@ bool GridFile::get(Mesh *m, Field2D &var,   const std::string &name, BoutReal de
+  if (field_dimensions[0] == m->GlobalNx - 2*mxg ) {
+    for (int y=0; y<m->LocalNy; y++) {
+      for (int z=0; z<var.getNz(); z++) {
+        for (int x=0; x<m->xstart; x++) {
```
The loops over y- and z- are repeated, but it's only a couple of extra lines and probably looks less odd - I'll move the x-loops to the outside.

johnomotani commented:
Fixes #24.

johnomotani commented:
I've added a test case. It generates some grid files to test in the python `runtest` script, so no binary files committed :-) It might be a bit slow to run because the test needs to use 6 processors so that it can load a double-null topology.

I also found that the existing `test-gridfile` was checking the minimum absolute error instead of the maximum. I've fixed it and it still passes.

ZedThree commented:
Thanks @johnomotani -- tests look nice and comprehensive! Is `make_inputs.py` needed? It looks like it does the same thing as `runtest`, without running the tests?

johnomotani commented:
Sorry, shouldn't have added `make_inputs.py`. I'd had it separate from `runtest` to start with and forgot to delete after I merged them.

johnomotani commented:
I think I've fixed the unit test fails. I've made `OptionsTest` use the `FakeMeshFixture`. I'm not quite sure what it was doing before. In this PR there was a segfault when `source->hasXBoundaryGuards(this)` was called in the `Coordinates` constructor. I'm a bit surprised the test was managing to construct `Coordinates` objects in `getCoordinates()` successfully before... It would be a bit neater if it was possible to just call `fieldmesh.setCoordinates(nullptr)` to get around this, but if I just do that, `create3D` fails because it tries to use `fromFieldAligned` and there is no `ParallelTransform` available. The nice solution to that would be to set `Options::root()["input"]["transform_from_field_aligned"] = false;` - this doesn't work because the `FieldFactory` is a singleton which is usually created already in a different test, so doesn't read the option. This option is set in `test_solver.cxx` but unless I'm confused, it won't be having any (reliable) effect there either? [note setting the option does work if, eg, you only run one test so that the `FieldFactory` singleton isn't already created before it's set.]

johnomotani commented:
WIP because I want to set the boundary cells of some more fields: Christoffel symbols, etc. Update coming soon...

johnomotani commented:
I've updated so that Christoffel symbol terms get extrapolated into boundary guard cells. I have a situation where I want to invert a Laplacian at the upper y-boundary position (`location==CELL_YLOW`, `jy=mesh->yend+1`) which was failing because `G1` and `G3` were NaN there. Extrapolating isn't the nicest solution, because the upper and lower boundary points are not treated the same for staggered fields - on the lower boundary (`location==CELL_YLOW`, `jy=mesh->ystart`) we can take derivatives normally so don't need to extrapolate. It would be a massive faff to do anything fancier than extrapolation though, so I think this update is better than nothing.
